emop-maven-plugin/src/main/java/edu/cornell/VmsMojo.java

```diff
@@ -48,103 +96,239 @@ public class VmsMojo extends DiffMojo {
+        }
+
+        saveViolationCounts();
+        if (!showAllInFile) {
+            rewriteViolationCounts();
+        }
+    }
+
+    private static class SurefireOutputHandler implements InvocationOutputHandler {
+        private static enum State {
+            PROLOGUE,   // Output preceding the surefire:test goal
+            BODY,       // Output from surefire:test
+            EPILOGUE    // Output following the surefire:test goal
+        }
+
+        private static final Pattern epilogueStart = Pattern.compile("\\[INFO\\] BUILD (SUCCESS|FAILURE).*");
+
+        private ViolationFilterer filterer;
+        private State state = State.PROLOGUE;
+        private boolean filteringCurrentViolation = false;
+
+        public SurefireOutputHandler(ViolationFilterer filterer) {
+            this.filterer = filterer;
+        }
+
+        @Override
+        @SuppressWarnings("checkstyle:Regexp")
+        public void consumeLine(String line) {
+            switch (state) {
+                case PROLOGUE:
+                    if (line.contains("maven-surefire-plugin")) {
+                        System.out.println(line);
+                        state = State.BODY;
+                    }
+                    break;
+                case BODY:
+                    if (line.startsWith("Specification")) {
+                        if (!(filteringCurrentViolation = filterer.filter(line))) {
+                            System.out.println(line);
+                        }
+                    } else if (epilogueStart.matcher(line).matches()) {
+                        state = State.EPILOGUE;
+                    } else if (line.startsWith("[INFO]") || !filteringCurrentViolation) {
+                        System.out.println(line);
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+    }
+
+    private class ViolationFilterer {
+        private final boolean filterOld = !firstRun && !showAllInConsole;
+
+        /**
+         * Returns whether to filter the violation.
+         *
+         * @param violation A string representation of a violation
+         * @return <code>true</code> if the violation should be filtered; <code>false</code> otherwise
+         */
+        private boolean filter(String violation) {
+            if (filterOld) {
+                Violation newViolation = Violation.parseViolation(violation);
+                for (Violation oldViolation : oldViolations) {
+                    if (isSameViolationAfterDifferences(oldViolation, newViolation)) {
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+    }
+
+    /**
+     * Runs Maven Surefire.
+     *
+     * @throws MojoExecutionException if the Surefire execution fails.
+     */
+    private void invokeSurefire() throws MojoExecutionException {
```

Comment:
I didn't know you can invoke maven surefire like this, this is very cool!! Thanks for teaching me something new :)